mat64/eigen.go

```diff
@@ -798,22 +800,23 @@ func hqr2(d, e []float64, hess, v *Dense, epsilon float64) {
 	}
 }
 
-// D returns the block diagonal eigenvalue matrix from the real and imaginary
-// components d and e.
-func (f EigenFactors) D() *Dense {
-	d, e := f.d, f.e
-	var n int
-	if n = len(d); n != len(e) {
-		panic(ErrSquare)
-	}
-	dm := NewDense(n, n, nil)
-	for i := 0; i < n; i++ {
-		dm.set(i, i, d[i])
-		if e[i] > 0 {
-			dm.set(i, i+1, e[i])
-		} else if e[i] < 0 {
-			dm.set(i, i-1, e[i])
+type EigenValues []complex128
+
+// D returns the eigenvalue matrix from the real and imaginary components
+// of the eigenvalues.
+//
+// If the decomposed matrix is not symmetric, then D returns a block diagonal
+// matrix with the real eigenvalues in 1-by-1 blocks and any complex eigenvalues,
+// lambda + i*mu, in 2-by-2 blocks, [lambda, mu; -mu, lambda].
+func (v EigenValues) D() *Dense {
```

Comment:
The example you sent me when we were discussing implementing Pow shows this, it is at https://github.com/gonum/matrix/pull/83/files#diff-10d11ada769df6b8c836f6a142f54d52R90. The eigenvalues are correct, but the eigenvectors differ:

```
> a = [1,6,7; 9,1,10; 15,6,1]
a =

    1    6    7
    9    1   10
   15    6    1

> [v,d] = eig(a)
v =

   0.47304 + 0.00000i  -0.14712 + 0.14441i  -0.14712 - 0.14441i
   0.61606 + 0.00000i  -0.63800 - 0.23295i  -0.63800 + 0.23295i
   0.62984 + 0.00000i   0.70441 + 0.00000i   0.70441 - 0.00000i

d =

Diagonal Matrix

   18.1344 +  0.0000i                    0                    0
                    0   -7.5672 +  1.0909i                    0
                    0                    0   -7.5672 -  1.0909i
```

equivalent from mat64:
```
⎡0.47304093468892405  -0.6629102254940207   0.6029367060177389⎤
⎢ 0.6160598119673665  -2.7326877058673555   -1.117348566292792⎥
⎣ 0.6298432997082382    3.059661412350093  0.11648819407649919⎦

mat64.EigenValues{(18.13437754618839+0i), (-7.567188773094204+1.0909493212308445i), (-7.567188773094204-1.0909493212308445i)}
```

The issue is recovering (and representing) the conjugates that matlab shows.